### ISSUE TITLE
All tree nodes should have a fileOffset different from -1

### ISSUE THREAD

**Issue by peter-ahe-google** [dart-lang/sdk#32289]
I believe that all kernel nodes that Fasta generates should have a fileOffset that isn't -1. However, we have a number of tests that are currently asserting otherwise.

I believe that the property "isSynthetic" is different from "fileOffset" and I believe that it is important that even synthetic elements point to somewhere relevant.

A few examples:

* default constructors can point to the class name.
* implicit initializers (null literals) can point to what they initialize.
* Int literals from ++ and -- can point to the operator.

**Comment by jensjoha** [comment]
If i recall correctly, in the VM, isSynthetic basically just checks if the token position is < some constant (lets say -20) --- and can be converted to a non-synthetic so that -30 would be 10 or something similar... Whether or not that's reported I'm not sure...

**Comment by peter-ahe-google** [comment]
That's interesting. I was actually contemplating using negative numbers to signal synthetic nodes. This is appealing to me as it reduces the need for a field on TreeNode to track the synthetic bit.

**Comment by danrubel** [comment]
In the scanner and parsers (both fasta and analyzer), synthetic tokens may have a length of zero, but AFAIK all token offsets are >= 0.

**Comment by bwilkerson** [comment]
Correct. In analyzer, synthetic tokens have the offset of the token that would have followed them had they existed in the source code. For example, given `if () {}`, we will insert a synthetic token as the condition, and the offset will be the offset of the right parenthesis.

This is actually a requirement because code completion needs to be able to find the token by offset in order to know what suggestions to make.

**Comment by rmacnak-google** [comment]
The VM uses negative token positions to denote synthetic source positions, -1 to denote no source (0 is a valid token position for us), and positive for regular source positions. The debugger only stops at regular source positions, but profile ticks apply to regular or synthetic source positions.

https://github.com/dart-lang/sdk/blob/75bed68ca2bf63aef70b1e7b5c58ff5f76ad19eb/runtime/vm/token_position.h#L13
